lib/rules/require-context-role.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+
+const AstNodeInfo = require('../helpers/ast-node-info');
+const Rule = require('./base');
+
+const ROOTS = {
+  row: ['rowheader', 'columnheader', 'gridcell'],
+  group: ['listitem', 'menuitem', 'menuitemradio', 'treeitem'],
+  menu: ['menuitem', 'menuitemcheckbox', 'menuitemradio'],
+  menubar: ['menuitem', 'menuitemcheckbox', 'menuitemradio'],
+  list: ['listitem'],
+  listbox: ['option'],
+  grid: ['row', 'rowgroup'],
+  rowgroup: ['row'],
+  treegrid: ['row'],
+  tablist: ['tab'],
+  tree: ['treeitem'],
+};
+
+const CHILDREN_FOR_ANALYZE = new Set(
+  Object.keys(ROOTS).reduce((acc, keyName) => acc.concat(ROOTS[keyName]), [])
+);
+const VALID_PARENTS_FOR_CHILD = Object.keys(ROOTS).reduce((result, keyName) => {
+  const children = ROOTS[keyName];
+  children.forEach((childName) => {
+    if (!(childName in result)) {
+      result[childName] = [];
+    }
+    result[childName].push(keyName);
+  });
+  return result;
+}, {});
+Object.keys(VALID_PARENTS_FOR_CHILD).forEach((key) => {
+  VALID_PARENTS_FOR_CHILD[key].sort();
+});
+
+function errorMessage(role) {
+  const roles = VALID_PARENTS_FOR_CHILD[role] || [];
+  return `You have an element with the role of "${role}" but it is missing the required (immediate) parent element of "[${roles.join(
+    ', '
+  )}]". Reference: https://www.w3.org/TR/wai-aria-1.0/roles#${role}.`;
+}
+
+const metaForNode = new WeakMap();
+
+module.exports = class RequireContextRole extends Rule {
+  visitor() {
+    return {
+      ElementNode(node) {
```

Comment:
Can we refactor this use the path ancestors stuff instead?

Reply:
done